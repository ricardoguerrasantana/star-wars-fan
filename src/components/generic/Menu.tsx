// React 
import { ReactNode } from 'react';

// debugger
import Debug from "debug";
const log = Debug("App:Menu");
log.log = console.log.bind(console);

// Prop types
export type MenuItem = {
  class: string;
  clickable?: boolean;
  component: ReactNode;
  handleClick?: () => void;
  id: string;
}

export type Props = {
  menuStyles: {
    container: string;
  };
  items: MenuItem[];
}

/** Menu generic component is a box that renders an array of boxes called Items that in turn receive custom components. These Items can be assigned handler functions from parent component. */
function Menu({ items , menuStyles }: Props) {
  log('Rendering...');
  return (
    <div className={menuStyles.container} >
      {items.map((item) => {
        return (
          <div 
            className={item.class}
            key={item.id}
            onClick={item.handleClick}
            style={item.clickable ? {cursor: "pointer"} : undefined}
          >
            {item.component}
          </div>
        );
      })}
    </div>
  );
}

export default Menu;