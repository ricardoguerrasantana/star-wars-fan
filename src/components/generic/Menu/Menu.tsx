// React 
import { ReactNode } from 'react';

// debugger
import Debug from "debug";
const log = Debug("App:Menu");
log.log = console.log.bind(console);

// Prop types
export type MenuItem = {
  element: ReactNode;
  handleClick?: () => void;
  id: string;
  pointer?: boolean;
  styles: string;
}

export type Props = {
  styles: {
    container: string;
  };
  items: MenuItem[];
}

/** Menu generic component is a box that renders an array of 
 * boxes called Items that in turn receive custom components. 
 * These Items can be assigned handler functions from parent 
 * component. */
function Menu({ items , styles }: Props) {
  log("Rendering...");
  return (
    <div 
      className={styles.container}
      data-testid="Menu" 
    >
      {items.map((item, i) => {
        return (
          <div 
            className={item.styles}
            data-testid="item"
            key={item.id}
            onClick={item.handleClick}
            style={item.pointer ? {cursor: "pointer"} : undefined}
          >
            {item.element}
          </div>
        );
      })}
    </div>
  );
}

export default Menu;