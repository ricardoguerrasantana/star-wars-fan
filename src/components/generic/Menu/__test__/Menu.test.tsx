// Testing libraries
import { render, screen, fireEvent, within } from '@testing-library/react';
import renderer from "react-test-renderer";

// Component
import Menu from '../Menu';

describe("<Menu />", () => {
  /** Fake props to be passed down to <Menu /> */
  const handleClick = jest.fn();
  /** Defining itmes array. */
  const Button = (
    <button 
      data-testid="button" 
      type="button" 
    />
  );
  const item = {
    element: Button,
    handleClick: handleClick,
    styles: "itemStyles",
  };
  const items = [
    {...item , id: "1", pointer: false},
    {...item , id: "2", pointer: true},
    {...item , id: "3", pointer: false},
    {...item , id: "4", pointer: true},
    {...item , id: "5", pointer: false},
  ]
  /** Menu styles. */
  const styles = {container: "containerStyles"}
  /** Number of click fires on item. */
  const clickTimes = 3;

  const MockMenu = () => (
    <Menu 
      items={items}
      styles={styles}
    />
  );
    
  it("should render menu component", () => {
    render(<MockMenu />);
    // screen.debug();
    
    expect(screen.getByTestId("Menu")).toBeInTheDocument();
  });
      
  it(`should render ${items.length} menu items.` , () => {
    render(<MockMenu />);
    const itemElements = screen.getAllByTestId("item");
    expect(itemElements.length).toBe(items.length);
  });
    
  it(`Menu items should be able to render ${items.length} elements each within an item.`, () => {
    const { getAllByTestId } = render(<MockMenu />);
    const itemElements = getAllByTestId("item");
    // Checks whether the expected element was inserted within each item.
    itemElements.forEach((element) => {
      expect(within(element).getByTestId("button")).toBeTruthy();
    });
  });
    
  it(`should click each of the ${items.length} button components ${clickTimes} times each.`, () => {
    const { getAllByTestId } = render(<MockMenu />);
    const itemElements = getAllByTestId("item");
    // Mocks clicking several times an item in menu.
    itemElements.forEach((item) => {
      for (let i = 0; i < clickTimes; i++) {
        fireEvent.click(item);
      }
    }); 
    expect(handleClick).toBeCalledTimes(itemElements.length * clickTimes);
  });


  it(`should test pointer is able to be set up on items.`, () => {
    const { getAllByTestId } = render(<MockMenu />);
    const itemElements = getAllByTestId("item");
    // Checks for defined cursor pointer style items
    expect(itemElements[0]).not.toHaveStyle("cursor: pointer;");
    expect(itemElements[1]).toHaveStyle("cursor: pointer;");
    expect(itemElements[2]).not.toHaveStyle("cursor: pointer;");
    expect(itemElements[3]).toHaveStyle("cursor: pointer;");
    expect(itemElements[4]).not.toHaveStyle("cursor: pointer;");
  });

  it(`should match the snapshot of initial render`, () => {
      const tree = renderer.create(<MockMenu />, ).toJSON();
      expect(tree).toMatchSnapshot();
  });
});