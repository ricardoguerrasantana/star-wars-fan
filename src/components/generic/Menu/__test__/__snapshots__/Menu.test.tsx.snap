// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> matches snapshot 1`] = `
<div
  className="containerStyles"
  data-testid="Menu"
>
  <div
    className="itemStyles"
    data-testid="item"
    onClick={[MockFunction]}
  >
    <button
      data-testid="button"
      type="button"
    />
  </div>
  <div
    className="itemStyles"
    data-testid="item"
    onClick={[MockFunction]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    <button
      data-testid="button"
      type="button"
    />
  </div>
  <div
    className="itemStyles"
    data-testid="item"
    onClick={[MockFunction]}
  >
    <button
      data-testid="button"
      type="button"
    />
  </div>
  <div
    className="itemStyles"
    data-testid="item"
    onClick={[MockFunction]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    <button
      data-testid="button"
      type="button"
    />
  </div>
  <div
    className="itemStyles"
    data-testid="item"
    onClick={[MockFunction]}
  >
    <button
      data-testid="button"
      type="button"
    />
  </div>
</div>
`;

exports[`<Menu /> should change cursor when hovered and able to be clicked 1`] = `
<div
  className="containerStyles"
  data-testid="Menu"
>
  <div
    className="itemStyles"
    data-testid="item"
    onClick={[MockFunction]}
  >
    <button
      data-testid="button"
      type="button"
    />
  </div>
  <div
    className="itemStyles"
    data-testid="item"
    onClick={[MockFunction]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    <button
      data-testid="button"
      type="button"
    />
  </div>
  <div
    className="itemStyles"
    data-testid="item"
    onClick={[MockFunction]}
  >
    <button
      data-testid="button"
      type="button"
    />
  </div>
  <div
    className="itemStyles"
    data-testid="item"
    onClick={[MockFunction]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    <button
      data-testid="button"
      type="button"
    />
  </div>
  <div
    className="itemStyles"
    data-testid="item"
    onClick={[MockFunction]}
  >
    <button
      data-testid="button"
      type="button"
    />
  </div>
</div>
`;
